packages/ts-transformers/src/tests/preserve-blank-lines-test.ts

```diff
@@ -0,0 +1,249 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {test} from 'uvu';
+import {compileTsFragment, CompilerHostCache} from './compile-ts-fragment.js';
+import * as ts from 'typescript';
+import * as assert from 'uvu/assert';
+import * as prettier from 'prettier';
+import preserveBlankLinesTransformer from '../preserve-blank-lines.js';
+
+const cache = new CompilerHostCache();
+
+/**
+ * Compile the given fragment of TypeScript source code using
+ * preserveBlankLinesTransformer. Check that there are no errors and that the
+ * output matches (prettier-formatted).
+ */
+function checkTransform(inputTs: string, expectedJs: string) {
+  const options = ts.getDefaultCompilerOptions();
+  options.target = ts.ScriptTarget.ESNext;
+  options.module = ts.ModuleKind.ESNext;
+  // Don't automatically load typings from nodes_modules/@types, we're not using
+  // them here, so it's a waste of time.
+  options.typeRoots = [];
+  const result = compileTsFragment(inputTs, __dirname, options, cache, () => ({
+    before: [preserveBlankLinesTransformer()],
+  }));
+
+  let formattedExpected = prettier.format(expectedJs, {parser: 'typescript'});
+  // TypeScript >= 4 will add an empty export statement if there are no imports
+  // or exports to ensure this is a module. We don't care about checking this.
+  const unformattedActual = (result.code || '').replace('export {};', '');
+  let formattedActual;
+  try {
+    formattedActual = prettier.format(unformattedActual, {
+      parser: 'typescript',
+    });
+  } catch {
+    // We might emit invalid TypeScript in a failing test. Rather than fail with
+    // a Prettier parse exception, it's more useful to see a diff.
+    formattedExpected = expectedJs;
+    formattedActual = unformattedActual;
+  }
+  assert.is(formattedActual, formattedExpected);
+  assert.equal(
+    result.diagnostics.map((diagnostic) =>
+      ts.formatDiagnostic(diagnostic, result.host)
+    ),
+    []
+  );
+}
+
+test('empty file', () => {
+  const input = ``;
+  const expected = ``;
+  checkTransform(input, expected);
+});
+
+test('no blank lines', () => {
+  const input = `
+    import 'foo';
+    class Foo {
+      foo() {}
+      bar() {}
+    }
+  `;
+  const expected = `
+    import 'foo';
+    class Foo {
+      foo() {}
+      bar() {}
+    }
+  `;
+  checkTransform(input, expected);
+});
+
+test('blank line after imports', () => {
```

Comment:
Is it possible for there to be a file with only new lines, or begins with newlines before the code?

Reply:
Does the transforming code handle syntax errors, or is that not applicable?

Thank you! This code change is fantastic!

Reply:
> Is it possible for there to be a file with only new lines, or begins with newlines before the code?

Good questions! Investigated this, and found that comments at the beginning and end of the file are weird!

Comments at the beginning of the file are associated with the `SourceFile` *and* the first child node. Comments at the end of the file are associated with the `EndOfFileToken`. In both cases, it is not possible to change the default comment-emitting behavior (which drops blank lines) -- you can't use the `setTextRange` trick, and you also can't use `addSyntheticComment` (they have no effect). So unless we can find another trick, there isn't much we can do here. I'm not too worried, since dropping blank lines at the very beginning and end of the file doesn't seem like that big a deal. Added some `.skip` tests with a `TODO` to document this in case we want to revisit it later.

This did help me find and fix a bug that affected an important case, though, which is something like:

```ts
/** @license ... */

const foo = 0;
```

While there is a comment at the top of the file here, so we couldn't deal with any blank lines *above* it -- we can still insert a newline above the `const foo = 0;` to recover *that* blank line. We were sort of doing this before, but in a way that actually caused the leading comment to be emitted twice. Fixed now with a test.

> Does the transforming code handle syntax errors, or is that not applicable?

We shouldn't need to do anything special. If TypeScript encounters a syntax error, it will try its best to produce a valid AST, which is what we'll transform, and the user will see a syntax error.